external/pkg/docker-exec-ws/shellclient.go

```diff
@@ -0,0 +1,353 @@
+package dockerExecWS
+
+import (
+	"encoding/binary"
+	"fmt"
+	"io"
+	"net/url"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/taskcluster/taskcluster-cli/external/pkg/iopipes"
+	"github.com/taskcluster/taskcluster-worker/plugins/interactive/shellconsts"
+	"github.com/taskcluster/taskcluster-worker/runtime/atomics"
+)
+
+const (
+	// MaxOutstandingBytes is the maximum number of unsent bytes in the stdin buffer.
+	MaxOutstandingBytes = 8 * 1024 * 1024
+	pingInterval        = 15 * time.Second
+	writeTimeout        = 2 * pingInterval
+	readTimeout         = 3 * pingInterval
+)
+
+var dialer = websocket.Dialer{
+	HandshakeTimeout: shellconsts.ShellHandshakeTimeout,
+	ReadBufferSize:   shellconsts.ShellMaxMessageSize,
+	WriteBufferSize:  shellconsts.ShellMaxMessageSize,
+}
+
+// ShellClient is the implementation of
+// github.com/taskcluster/taskcluster-worker/engines.Shell for v1 shells.
+type ShellClient struct {
+	ws       *websocket.Conn
+	done     chan struct{}
+	mWrite   sync.Mutex
+	cPause   *sync.Cond
+	paused   bool
+	complete atomics.Once
+	success  bool
+	err      error
+
+	stdin        io.WriteCloser
+	stdinReader  io.ReadCloser
+	stdout       io.ReadCloser
+	stdoutWriter *iopipes.DrainingPipeWriter
+	stderr       io.ReadCloser
+	stderrWriter *iopipes.DrainingPipeWriter
+}
+
+// Dial will built a proper shell websocket URL, connect to the server and return an initialized shell client.
+func Dial(socketURL string, command []string, tty bool) (*ShellClient, error) {
```

Comment:
Maybe I suggest that we add an additional optional argument `logger`... like:

```go
// Logger defines the interface for writing log messages, implemented by logrus and others..
type Logger interface {
  Debugf(format string, a ...interface{})
  Infof(format string, a ...interface{})
  Warnf(format string, a ...interface{})
  Errorf(format string, a ...interface{})
}

// A trivial logger implementation that discards all messages
type nilLogger struct{}
func (nilLogger) Debugf(string, ...interface{}) {}
func (nilLogger) Infof(string, ...interface{}) {}
func (nilLogger) Warnf(string, ...interface{}) {}
func (nilLogger) Errorf(string, ...interface{}) {}

//.... logger is optional
func Dial(socketURL string, command []string, tty bool, logger Logger) (*ShellClient, error) {
  if logger == nil {
    logger = nilLogger{}
  }

}
```
Now instead of littering the code with `fmt.Fprintln(os.Stderr, "Resuming client")` which is nice for debugging, we can specify a logger which we use during tests and in development... and when we do production we just give `nil` as logger and messages go away...

We could even implement a verbose mode in the tool like: `taskcluster shell $TASKID --verbose [--] [command...]`, which would print some or all of these warning messages...

Note: the way I proposed the `Logger` interface it is already implemented by https://godoc.org/github.com/sirupsen/logrus
 and by https://github.com/taskcluster/taskcluster-worker/blob/79da11e34f8b2e7ef8a01c101749717e00b1666b/runtime/monitor.go
which is used in tc-worker... This is kind of a neat feature with golang...